posts/ngrx-facade-pattern.md

```diff
@@ -0,0 +1,137 @@
+
+### What is a Facade?
+In software, a facade provides a public interface to mask the usage of something more complex. In this instance, it will neatly wrap our NgRx interactions and allow one point of contact for our component and state management without our component having any knowledge of NgRx.
+
+### Implementation
+Before jumping in to the facade example, here is a component using NgRx without a facade.
+``` Typescript
+
+export class AccountListComponent implements OnInit {
+  accountList: Observable<Account[]>;
+
+  constructor(private store: Store<AccountListState>) {}
+
+  ngOnInit() {
+    this.accountList = this.store.select(getAccountList);
+
+    this.loadAccountList();
+  }
+
+  loadAccountList() {
+    this.store.dispatch(new LoadAccountList());
+  }
+
+  addAccount(account: string) {
+    this.store.dispatch(new AddAccount(account));
+  }
+
+  editAccount(id: string, account: string) {
+    this.store.dispatch(new EditAccount({ id, account }));
+  }
+
+  deleteAccount(id: string) {
+    this.store.dispatch(new DeleteAccount(id));
+  }
+}
+
+```
+
+This implementation is very `store` dependent and makes the component more complex. By adding a facade for the component to interact with, we can simplify the access to our store and decouple the knowledge of NgRx in our component. By performing that separation of duties, if we choose to remove NgRx in the future we do not have to refactor our component.
+
+Now let's create an `AccountService` that will act as our facade for NgRx.
+``` Typescript
+
+@Injectable({
+  providedIn: 'root',
+})
+export class AccountService {
+  accountList = this.store.select(getAccountList);
+
+  constructor(private store: Store<AccountListState>) {}
+
+  loadAccountList() {
+    this.store.dispatch(new LoadAccountList());
+  }
+
+  addAccount(account: string) {
+    this.store.dispatch(new AddAccount(account));
+  }
+
+  editAccount(id: string, account: string) {
+    this.store.dispatch(new EditAccount({ id, account }));
+  }
+
+  deleteAccount(id: string) {
+    this.store.dispatch(new DeleteAccount(id));
+  }
+}
+
+```
+
+After moving everything related to the store into our service, we can then inject that service into our component.
+
+``` Typescript
+
+export class AccountListComponent implements OnInit {
+  accountList: Observable<Account[]>;
+
+  constructor(private accountService: AccountService) {}
+
+  ngOnInit() {
+    this.accountList = this.accountService.accountList;
+
+    this.accountService.loadAccountList();
+  }
+
+  addAccount(account: string) {
+    this.accountService.addAccount(account);
+  }
+
+  editAccount(id: string, account: string) {
+    this.accountService.editAccount( id, account);
+  }
+
+  deleteAccount(id: string) {
+    this.accountService.deleteAccount(id);
+  }
+}
+
+```
+After implementing the facade pattern what did we actually gain?
+
+### Pros
+- Better developer experience.
+    - Devs only interact with a service instead of NgRx stores, actions, and reducers.
```

Comment:
Presumably some dev will also have to write the AccountService and deal with NgRx stuff... Maybe "Devs have clearer focus on either NgRx interactions or component actions, without having to think about both at the same time" or something like that?

Reply:
updated